components/script/dom/globalscope.rs

```diff
@@ -207,6 +378,181 @@ impl GlobalScope {
+    /// we wait two turns of the event-loop,
+    /// to ensure any tasks left in the queue are handled.
+    /// No new messages/tasks can be enqueued while this is ongoing(detached is set).
+    ///
+    /// https://html.spec.whatwg.org/multipage/#ports-and-garbage-collection
+    pub fn perform_a_message_port_garbage_collection_checkpoint(&self) {
+        for (port_id, counter) in self.message_ports_to_be_collected.borrow_mut().iter_mut() {
+            *counter += 1;
+            if *counter > 1 {
+                let port = self
+                    .message_ports
+                    .borrow_mut()
+                    .remove(port_id)
+                    .expect("GC'ed port to exist");
+                self.message_ports_shipped.borrow_mut().remove(port_id);
+                if port.detached() {
+                    let _ = self
+                        .script_to_constellation_chan()
+                        .send(ScriptMsg::RemoveMessagePort(*port_id));
+                }
+            }
+        }
+
+        self.message_ports_to_be_collected
+            .borrow_mut()
+            .retain(|_port_id, counter| *counter < 2);
+
+        // Look for any ports that have been detached.
+        for (id, port) in self.message_ports.borrow().iter() {
+            if port.detached() || port.transferred() {
+                self.start_garbage_collecting_message_port(&id);
+            }
+        }
+    }
+
+    /// Start tracking a message-port:
+    ///
+    /// 1. Add port to pending_message_ports(avoiding a double-borrow on message_ports).
+    /// 2. Send an ipc-sender to the constellation,
+    ///    which will be used both to communicate the ipc-sender of the entangled port,
+    ///    when/if it ships as well,
+    ///    or to let this scope know when the entangled port has been dropped,
+    ///    The sender will also act as a way for the entangled port to send messages directly to here.
+    /// 3. Setup a route with a MessageListner to act as go-between for the ipc and the event-loop.
+    pub fn track_message_port(&self, port: &DomRoot<MessagePort>, transferred: bool) {
+        let message_port_id = port.message_port_id().clone();
+        if self
+            .message_ports_shipped
+            .borrow_mut()
+            .remove(&message_port_id)
+            .is_some()
+        {
+            self.message_ports_to_be_collected
+                .borrow_mut()
+                .remove(&message_port_id);
+        }
+
+        self.pending_message_ports
+            .borrow_mut()
+            .insert(message_port_id.clone(), port);
+
+        let (port_control_sender, port_control_receiver) =
+            ipc::channel().expect("ipc channel failure");
+
+        if !transferred {
+            let _ = self
+                .script_to_constellation_chan()
+                .send(ScriptMsg::NewMessagePort(
+                    message_port_id.clone(),
+                    port_control_sender.clone(),
+                ));
+        }
```

Comment:
Note: the first time a port is created, we must let the constellation know about it, hence the `transferred` argument. If the port was not newly created, but instead transfer received, we wait one task and only let the constellation know about the "new" port if it hasn't been transferred by then. (see the task queued just below, a  micro-task could do instead).